python/mxnet/rnn/rnn_cell.py

```diff
@@ -440,6 +440,153 @@ def __call__(self, inputs, states):
+
+        Returns
+        -------
+        args : dict of str -> NDArray
+            dictionary with weights associated to
+            this cell packed.
+        """
+        args = args.copy()
+        out_names = {'i2h': ['_o', '_z', '_r'],
+                     'h2h_gates': ['_z', '_r'],
+                     'h2h': ['_o']}
+
+        for group_name in out_names:
+            weight = []
+            bias = []
+            for name in out_names[group_name]:
+                wname = '%s%s%s_weight'%(self._prefix, group_name, name)
+                weight.append(args.pop(wname))
+                bname = '%s%s%s_bias'%(self._prefix, group_name, name)
+                bias.append(args.pop(bname))
+            args['%s%s_weight'%(self._prefix, group_name)] = ndarray.concatenate(weight)
+            args['%s%s_bias'%(self._prefix, group_name)] = ndarray.concatenate(bias)
+        return args
+
+    def __call__(self, inputs, states):
+        """Construct symbol for one step of RNN.
+
+        Parameters
+        ----------
+        inputs : sym.Variable
+            input symbol, 2D, batch * num_units
+        states : sym.Variable
+            state from previous step or begin_state().
+
+        Returns
+        -------
+        output : Symbol
+            output symbol
+        states : Symbol
+            state to next step of RNN.
+        """
+        self._counter += 1
+
+        seq_idx = self._counter
+        name = '%st%d_' % (self._prefix, seq_idx)
+        prev_state_h = states[0]
+
+        i2h = symbol.FullyConnected(data=inputs,
+                                    weight=self._i2h_weight,
+                                    bias=self._i2h_bias,
+                                    num_hidden=self._num_hidden * 3,
+                                    name="%s_i2h" % name)
+        h2h = symbol.FullyConnected(data=prev_state_h,
+                                    weight=self._h2h_gates_weight,
+                                    bias=self._h2h_gates_bias,
+                                    num_hidden=self._num_hidden * 2,
+                                    name="%s_h2h_gates" % name)
+        i2h, i2h_z, i2h_r = symbol.SliceChannel(i2h, num_outputs=3,
+                                                name="%s_i2h_slice" % name)
+        h2h_z, h2h_r = symbol.SliceChannel(h2h, num_outputs=2,
+                                           name="%s_h2h_gates_slice" % name)
+
+        update_gate = symbol.Activation(i2h_z + h2h_z, act_type="sigmoid", name="%s_z_act" % name)
+        reset_gate = symbol.Activation(i2h_r + h2h_r, act_type="sigmoid", name="%s_r_act" % name)
+
+        next_h_tmp = symbol.FullyConnected(data=(reset_gate * prev_state_h),
+                                           weight=self._h2h_weight,
+                                           bias=self._h2h_bias,
+                                           num_hidden=self._num_hidden,
+                                           name="%s_h2h" % name)
```

Comment:
I wasn't aware that cudnn was actually implementing a slight variation on GRUs. Their variation actually simplifies things a bit as you can represent everything in two matrix multiplications instead of three (like LSTMs). I guess the question is whether compatibility with cudnn or staying true to the original GRU is more important.